os/board/rtl8730e: fix wifi events for bridge

jwei5 opened the pull request:
- Fix wifi events not being handled properly for bridge mode (trigger EVT_LEFT when STA received no beacon)
- Clean up unnecessary codes from previous bridge mode PR
Note:  Will be checking for any further changes to RTK layer, g_mode in rtk_netmgr.c may not be necessary

Commit:
os/board/rtl8730e: fix wifi events for bridge

- Fix wifi events not being handled properly for bridge mode (trigger EVT_LEFT when STA received no beacon)
- Clean up unnecessary codes from previous bridge mode PR
- Modify get_info and get_signal_quality to handle correct case when ENABLE_HOMELYNK is enabled or disabled

hk-gwak commented on os/board/rtl8730e/src/component/os/tizenrt/rtk_netmgr.c:
```diff
@@ -776,17 +720,11 @@ trwifi_result_e wifi_netmgr_utils_get_info(struct netdev *dev, trwifi_info *wifi
+			if (wifi_is_connected_to_ap() == RTK_STATUS_SUCCESS) {
+				rtw_phy_statistics_t phy_statistics;
+				if (wifi_fetch_phy_statistic(&phy_statistics) == RTK_STATUS_SUCCESS){
+					wifi_info->rssi = (int)phy_statistics.rssi;
```
Currently (before homelynk feature) we only have wlan0, and it can be softAP or STA mode, which is like situation that CONFIG_ENABLE_HOMELYNK == n. When CONFIG_ENABLE_HOMELYNK == n, I'm wondering if this PR can distinguish softAP from STA mode.

jwei5 replied:
I think g_mode is not suitable, I'm testing if it is better to use API to check wifi mode.
Currently, only `wifi_netmgr_utils_get_info `and `wifi_netmgr_utils_get_signal_quality `can be used by softAP mode.
Since only wlan0 is passed into these 2 functions, we can use API to check if softAP mode is running instead of `g_mode == RTK_WIFI_SOFT_AP_IF`

BTW, for `wifi_netmgr_utils_get_info`, do you still require softAP mode to return `wifi_info->wifi_status = TRWIFI_SOFTAP_MODE` ? I tried to find in TizenRT but it seems like `wifi_status` is not used anywhere

hk-gwak replied:
Yes, requirement is to return wifi_info->wifi_status = TRWIFI_SOFTAP_MODE when softAP mode. In TizenRT 5.0, we are still using this `wifi_state`.
For g_mode, let me think about this a liite bit more.

jwei5 commented on os/board/rtl8730e/src/component/os/tizenrt/rtk_netmgr.c:
```diff
@@ -776,17 +723,16 @@ trwifi_result_e wifi_netmgr_utils_get_info(struct netdev *dev, trwifi_info *wifi
 			wifi_info->rssi = (int)0;
-			if (g_mode == RTK_WIFI_SOFT_AP_IF) {
+#ifndef CONFIG_ENABLE_HOMELYNK
+			if (wifi_is_running(SOFTAP_WLAN_INDEX)) {
```
Hi @hk-gwak , added back the condition to return `wifi_info->wifi_status = TRWIFI_SOFTAP_MODE` in softAP mode.

Also proposed an implementation about distinguishing softAP from STA mode when `CONFIG_ENABLE_HOMELYNK == n` in this amended commit:
- Even with bridge mode enabled, I am assuming that only `wlan0 `will call into `utils_get_info()` and `utils_get_signal_quality()`
- If CONFIG_ENABLE_HOMELYNK == n, we check if softAP is running in driver first, before we do the flow for STA mode.
- If wlan1 will also call into these functions, then we may need to add condition to check `dev->ifname`

Do let me know if there are any issues

hk-gwak replied:
Thank you for your explanation. I don't know the rtk driver's structure well.... so it was a little bit hard to understand.
Actually in TizenRT 5.0, not only TRWIFI_SOFTAP_MODE but also TRWIFI_CONNECTED / TRWIFI_DISCONNECTED values are used(those values are for when CONFIG_ENABLE_HOMELYNK == n). We still need those values by wifi_netmgr_utils_get_info(). In this PR, looks like I can get only TRWIFI_SOFTAP_MODE by wifi_netmgr_utils_get_info().
And is it possible to get value for knowing whether bridge mode is on by wifi_netmgr_utils_get_info()? Actually https://github.com/Samsung/TizenRT/pull/6851#discussion_r2137657099 this comment was for that value(bridge mode is on).

jwei5 replied:
Sorry about that, will add back `TRWIFI_CONNECTED` / `TRWIFI_DISCONNECTED` into `wifi_netmgr_utils_get_info()`
Just sharing some info about rtk driver for better understanding about the change:
- In our driver, STA is always using index 0, SoftAP is always using index 1, that's why we use `wifi_is_running()` to check for softAP mode even if there is only wlan0 in TizenRT (`CONFIG_ENABLE_HOMELYNK == n`)

Could I check what conditions will determine `TRWIFI_BRIDGE_MODE`? Must wlan0 be connected and wlan1 in softAP to be considered bridge mode?

hk-gwak replied:
In my opinion, TRWIFI_BRIDGE_MODE is hard to be determined by the state of wlan0 & wlan1 because there will be various circumstances during bridge mode is on. (`wlan0 be connected and wlan1 in softAP` is one of those circumstances. another circumstance is `wlan0 be disconnected and wlan1 in softAP`)

So let's just make one variable for bridge on/off, (ex. g_bridge_on)
bridge enable/disable will be delivered by wifi_netmgr_utils_set_bridge(). When bridge enabled, set g_bridge_on to true, and when bridge disabled, set g_bridge_on to false.
g_bridge_on == true -> TRWIFI_BRIDGE_MODE .

hk-gwak commented on os/board/rtl8730e/src/component/os/tizenrt/rtk_netmgr.c:
```diff
@@ -393,62 +373,39 @@ trwifi_result_e wifi_netmgr_utils_init(struct netdev *dev)
+		return wuret;
 	}
+	
+	/* Regardless of interface, set g_mode as STA during netmgr init as SoftAP is not running yet */
+	g_mode = RTK_WIFI_STATION_IF;
```
will be changed to RTK_WIFI_SOFT_AP_IF when start_softap?

hk-gwak commented on os/board/rtl8730e/src/component/os/tizenrt/rtk_netmgr.c:
```diff
@@ -368,17 +355,10 @@ static int rtk_drv_callback_handler(int type)
-		} else {
-			type = 7;
-		}
+	if (reason->if_id == RTK_WIFI_STATION_IF) {
+		type = 4; // STA disconnected
+	} else if (reason->if_id == RTK_WIFI_SOFT_AP_IF) {
+		type = 5; // STA left SoftAP
```
Each conditions are determined by reason->if_id, but reason->if_id is not changed when wifi_netmgr_utils_start_softap / wifi_netmgr_utils_stop_softap. Is this OK?

jwei5 replied:
Hi, you may refer to my comment on https://github.com/hk-gwak/TizenRT/pull/1/files#r2144860277 on how reason->if_id is used. Do let me know if you have any queries~

hk-gwak commented on os/board/rtl8730e/src/component/os/tizenrt/rtk_netmgr.c:
```diff
@@ -728,7 +672,8 @@ trwifi_result_e wifi_netmgr_utils_get_signal_quality(struct netdev *dev, trwifi_
 			u8 channel;
 			u32 tx_rty = 0;
-			if (g_mode == RTK_WIFI_SOFT_AP_IF) {
+#ifndef CONFIG_ENABLE_HOMELYNK
+			if (wifi_is_running(SOFTAP_WLAN_INDEX)) {
```
Looks like `wifi_is_running(SOFTAP_WLAN_INDEX)` is checking whether softAP is running or not. I can't understand why ifi_get_channel(), wifi_get_sw_statistic() is executed when softap is running when CONFIG_ENABLE_HOMELYNK == n. As far as I know, those functions are for when STA mode.

jwei5 replied:
When we were implementing `wifi_netmgr_utils_get_signal_quality()`, we were unsure if application side will be using this function as SoftAP, so we included support for it. 
If it is not needed, we can remove it.

hk-gwak replied:
OK, I understood. I think we 'd better keep this!

jwei5 commented on os/board/rtl8730e/src/component/wifi/inic/inic_ipc_host_api_basic.c:
```diff
@@ -362,6 +364,7 @@ static void wifi_ap_sta_assoc_hdl( char* buf, int buf_len, int flags, void* user
 	//USER TODO
 	rtk_reason_t reason;
 	memset(&reason, 0, sizeof(rtk_reason_t));
+	reason.if_id = RTK_WIFI_SOFT_AP_IF;
```
`reason.if_id` is only used to track which interface the wifi event is coming from. This reason is then checked in linkup/linkdown_handler in `rtk_netmgr.c`, and the correct wifi event will be passed into `wifi_manager`.
Whenever the `reason `is assigned, it will be memset to 0 first.
Example of usage:

1. Wifi driver receives connection from a client, it will trigger a callback into `wifi_ap_sta_assoc_hdl()`
2. As this event is received from SoftAP, `wifi_ap_sta_assoc_hdl()` will set the` reason.if_id = RTK_WIFI_SOFT_AP_IF`
3. It will pass this reason into `linkup_handler() `in `rtk_netmgr.c`
4. `linkup_handler()` will check `reason.if_id`. Since it is from SoftAP, it will indicate to `rtk_drv_callback_handler()` that the event to post to wifi_manager is `LWNL_EVT_SOFTAP_STA_JOINED`

hk-gwak commented:
Thank you for the explanation. Please just fix for the `TRWIFI_BRIDGE_MODE`.!

Commit:
os/board/rtl8730e: track bridge mode status and remove g_mode in rtk_netmgr.c

- Add flag g_bridge_on to track if bridge mode is running
- Modify wifi_netmgr_utils_get_info() to return TRWIFI_BRIDGE_MODE if bridge mode is running
- Remove use of g_mode as it is not necessary
- Add flag g_netmgr_init to track if wifi manager is initialized

jwei5 commented:
Modified wifi_netmgr_utils_get_info to return TRWIFI_BRIDGE_MODE
Also removed g_mode as it is not necessary. Added flag g_init_netmgr and use API to track net manager status instead

hk-gwak commented on os/board/rtl8730e/src/component/os/tizenrt/rtk_netmgr.c:
```diff
@@ -938,6 +927,11 @@ trwifi_result_e wifi_netmgr_utils_set_bridge(struct netdev *dev, uint8_t control
 //	ret = DO_SOMETHING(control);
```
I think this line can also be removed.
//	ret = DO_SOMETHING(control);

jwei5 replied:
Noted.
Since there is no further action needed for this API, I will remove the condition `if (ret == RTK_STATUS_SUCCESS)` as well

hk-gwak reviewed (approved):
LGTM
